PX4 Rover should have manual reverse

bo-rc opened the pull request:
I tried PX4 on a RPI3/Navio2 hardware, running rover configuration. I found myself frequently needed to back off the rover from a corner manually but the current code base does not support manual reverse for rovers.

I first tried to play around with the Mixer settings by changing the output offset for the throttle channel. The manual reverse is possible but this method disrupts many things. E.g. arming/disarming using sticks is possible but not safe, AUTO HOLD throttle level is no long at a neutral position, mission throttle, min and max throttle levels are changed etc.

This pull request implements rover with manual reverse support. It has the following behaviors:
* reverse is only possible in manual mode, the mid position of throttle channel (typically 1500  us) is the neutral or breaking, the lower half is for revere and the higher half is for going forward.
* arming requires throttle channel at neutral position when motor is not spinning. 
* auto mission uses only the forward range of the throttle channel. auto hold mode uses neutral/breaking throttle.
* for the users, the behavior of mission cruise settings stay the same. e.g. 50% cruise throttle will be mapped to the correct forward speed, i.e. 50% of full forward speed.
* Other vehicle configurations and behaviors not affected.

Tested using HIL GPS simulated by Vicon motion capture indoors:
(first, I use manual mode to drove the rover backward into position, then I switched to auto mission. There are 3 plain way points (marked on the floor) in the mission.)

![insta](https://user-images.githubusercontent.com/14265605/29998233-861972a2-8feb-11e7-9184-d8a069e364e0.gif)



I believe many people would find this implementation useful. Let me know what you think!

dagar commented on src/modules/gnd_att_control/GroundRoverAttitudeControl.hpp:
```diff
@@ -77,14 +78,15 @@ class GroundRoverAttitudeControl
+	int		_battery_status_sub{ -1};		/**< battery status subscription */
+	int		_ctrl_state_sub{ -1};		/**< control state subscription */
+	int		_manual_sub{ -1};			/**< notification of manual control updates */
+	int		_params_sub{ -1};			/**< notification of parameter updates */
```
Did you run this through astyle? If so what version are you using, this seems wrong?

bo-rc replied:
Yes, but I used 2.05.1. I will update my astyle version. 

It should be less annoying now. Thanks!

dagar commented:
Awesome! 

What it make sense to keep the rover mixers generic? I think adding reverse to the others isn't going to hurt them.

Do you have actual changes in commander, or just applying astyle? I'm working on a larger refactor of commander (https://github.com/PX4/Firmware/pull/7517), and once that's under control we can start enforcing astyle everywhere. Would you mind leaving the commander style changes out for now?

TSC21 commented:
@bo-rc do you mind updating the Devguide accordingly?

bo-rc commented:
@TSC21 Yes, I will update the Devguide.

@dagar I added a some amount of lines in commander.cpp for disarm/arm check for rovers. but the issue was mainly the formatting. I cleaned it up a bit, now it should be much less annoying.

Yes, I would like to keep the mixer generic. Do you know where in the code base should I look before the control signal reaches the mixer? I'm thinking to map thrust output for rover to [-1.0, 1.0] before it reaches the mixer in the control pipeline. 

Maybe the better solution is to add a parameter e.g. ESC_REVERSIBLE.

dagar commented:
Thanks, overall PR is much easier to read now.

I should try this on my own car. As I recall for mine you can't go straight into reverse, but need to pause at neutral first. How common are these details for car ESCs?

Do we actually need to worry about the non reversible case for rovers? I think it was simply ignored for the first pass at these simple controllers and the full range is all we need.

bo-rc commented:
@dagar It depends on the settings of the ESC. Most car ESCs support programming (using jumper pins for example) to set as one of Reverse/Forward, Reverse/Breaking/Forward, or Breaking/Forward only mode. So, this should be handled by the user on his/her hardware but we can document the process in the user's guide.

One trickier question is that what if a user uses a plane ESC for a rover, or the rover is a Nitro/Gas-powered car which typically cannot be set to have reverse.

I will test those and report back.

dagar commented:
I was mainly hoping to cover the most common case by default.

bo-rc commented:
With a programmable car ESC, everything works fine. 

If the ESC sets Reverse/Breaking/Forward mode, with Manual mode, the lower range does reverse, the mid point does breaking and the higher range does forward; you have to stop the car first to reverse motor spinning directions. If the ESC sets Breaking/Forward mode, with Manual mode only forward is possible; the mid position and lower does motor breaking. If the ESC sets Reverse/Forward mode, With Manual mode the car can jump from forward directly to Reverse is the stick is moved fast.

With Auto mission mode, only forward is possible and the speed is consistent.

It is an interesting case with a plane ESC. Because most ESCs calibrate the neutral position at powering up time, the range is automatically mapped to the higher range of the throttle stick. So, mid position and lower is neutral/stop, higher is forward. The car goes forward only in both manual and auto mission modes.

So, despite of the hardware details, the code runs as designed. Auto mission goes forward only. Manual has reverse if your ESC supports it.

TSC21 commented:
@bo-rc can you please rebase?

bo-rc commented:
@TSC21 Hey, It has been a long time and the code has diverged a lot. Rebasing will require a lot of work which I prefer not to deal with. I will simply close this pull request and submit a new one later.
